mc-sema/bin_descend/get_cfg.py

```diff
@@ -1086,10 +1112,142 @@ def processDataChunk(ea, size):
+def sign_extend(value, bits):
+    sign_bit = 1 << (bits - 1)
+    return (value & (sign_bit - 1)) - (value & sign_bit)
+
+def checkIfOffsetTable(ea):
+    """
+    Check if this is an offset table: that is, a table
+    off offsets that when added to table base result
+    in the VA of a jump target
+    """
+
+    DEBUG("LOOKING FOR OFFSET TABLE AT: {:08x}\n".format(ea))
+    # preconditions
+    # can only really do this when all sections are 
+    # correctly based
+    if not isLinkedElf():
+        DEBUG("\t... not a linked elf\n");
+        return False, 0, []
+
+    # only present in PIE executables
+    if not PIE_MODE:
+        DEBUG("\t... not in PIE mode\n");
+        return False
+
+    #TODO: revisit
+    if getBitness() != 64:
+        DEBUG("\t... not 64-bit\n");
+        return False, 0, []
+
+    # check that there is a code reference
+    # to ea somewhere
+    refs = [x for x in idautils.DataRefsTo(ea)]
+    if len(refs) == 0:
+        DEBUG("\t... no refs to ea\n");
+        return False, 0, []
+    
+    # assumes 32-bit offsets
+    # 1: EA + EA[n] = beginning of an instruction
+    entrycount = 0
+    entries = []
+    while True:
+        entry = readDword(ea+entrycount*4)
+        # no null entries
+        if entry == 0:
+            break
+
+        dest_guess = ea + sign_extend(entry, 64)
+        dest_guess &= 0xFFFFFFFFL
+        # has to point to code and to the
+        # start of an instruction
+        if isInternalCode(dest_guess) and isSaneReference(dest_guess):
+            DEBUG("\tAdded destination: {:08x}\n".format(dest_guess))
+            entries.append(dest_guess)
+            entrycount += 1
+        else:
+            DEBUG("\tInvalid destination: {:08x}\n".format(dest_guess))
+            # invalid entry
+            break
+
+    # minimum here is fairly arbitrary
+    return (entrycount > 1, entrycount, entries)
+
+def createOffsetTable(M, table_start, table_entries):
+
+    # create new OffsetTable message
+    OT = M.offset_tables.add()
+
+    # set table start
+    OT.start_addr = table_start
+
+    # loop through table_entries, and populate
+    # * original data (int32, readDword(table_start + i * 4))
+    # * point-to va (int64, table_entries[i])
+
+    for idx, entry in enumerate(table_entries):
+        orig_data = readDword(table_start + idx * 4)
+        # orig data value at table index
+        OT.table_offsets.append(orig_data)
+        # destination at that index
+        OT.destinations.append(entry)
+
+    jmp_refs = set()
+    for ref in idautils.DataRefsTo(table_start):
+        DEBUG("Checking ref to table...\n")
+        insn_t = idautils.DecodeInstruction(ref)
+        # check if REF points to LEA REG, <value>
+        if insn_t.itype == idaapi.NN_lea and insn_t.Operands[0].type == idc.o_reg:
+            DEBUG("Found a LEA\n")
+            dest_reg = idc.GetOpnd(ref, 0)
+
+            # get next 5 insts
+            cur_head = idc.NextHead(ref)
```

Comment:
Ah okay. I assumed that the only possible correct instruction would be an indirect jump, and the only possible place it could be is at the end of the block, because IDA will end the block on the `jmp`. What happens when it's not a `jmp`?